### PR TITLE
Alerting: Update rule access control to explicitly check for permissions "alert.rules:read" and "folders:read"

### PR DESCRIPTION
**Pull request by yuri-tceretian**
**What is this feature?**
This PR updates alert rule access control to check that the user has permission to read the folder and rules in the folder.

This change does not introduce any new requirements in GET API because currently, they use the authorization service that checks those permissions.
- "alert.rules:read" is checked here https://github.com/grafana/grafana/blob/7cec741baee6f038b09f6aea768fc3e526bbc8da/pkg/services/ngalert/api/authorization.go#L27-L28
- "folders:read" is checked by folder service https://github.com/grafana/grafana/blob/b0448b92e5cc2678eb38e8021128d9d059d67884/pkg/services/folder/folderimpl/folder.go#L148-L158

It changes the set of permissions required for DELETE and POST methods. Currently, they do not require `alert.rules:read` permission for the group to change it. This PR adds this requirement.

**Why do we need this feature?**
- This is needed to harden the authorization logic so any usage of the service would check all set of permissions
- This will be needed in provisioning API because some endpoints do not mention folder UID in the paths and do not use folder service to get the folder
- Since the migration of rule API to use folder UID instead of its title, this will let us remove some code that still fetches the folder just to check permissions.

**Which issue(s) does this PR fix?**:

Related to https://github.com/grafana/grafana/pull/77007

**Special notes for your reviewer:**

This change should not affect Grafana Enterprise users that use fixed roles `fixed:alerting.rules:writer` because it includes reader permissions.

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

# Release notice breaking change

The alert rule API methods now require more permissions for users to perform changes to rules. To create a new rule or update or delete an existing one, the user must have permission to read from the folder that stores the rules (i.e. permission `folder:read` in the scope of the rule's folder)  and permission to read alert rules in that folder (i.e. permission `alert.rules:read`

The standard roles already have all required permissions, and therefore, neither OSS nor Grafana Enterprise users who use the fixed roles (standard roles provided by Grafana) are affected. __Only Grafana Enterprise users who create custom roles can be affected__

### CODE REVIEW

**Review by JacobsonMT — APPROVED** [review]
LGTM, as discussed there might be some odd edge case gaps in the auth around rules:read this closes so documenting any would be great. Core logic looks sound though.

**Review by JacobsonMT — APPROVED** [review]
LGTM :rocket:

File: pkg/services/ngalert/accesscontrol/rules.go [hunk 1]
```diff
@@ -113,14 +139,16 @@ func (r *RuleService) AuthorizeAccessToRuleGroup(ctx context.Context, user ident
...
 		}
-	} else if len(change.Delete) > 0 {
-		// add a safeguard in the case of inconsistency. If user hit this then there is a bug in the calculating of changes struct
-		return fmt.Errorf("failed to authorize changes in rule group %s. Detected %d deletes but group was not provided", change.GroupKey.RuleGroup, len(change.Delete))
+	} else {
+		if len(change.Delete) > 0 {
+			// add a safeguard in the case of inconsistency. If user hit this then there is a bug in the calculating of changes struct
+			return fmt.Errorf("failed to authorize changes in rule group %s. Detected %d deletes but group was not provided", change.GroupKey.RuleGroup, len(change.Delete))
+		}
 	}
```
Comment (rwwiv) [inline pkg/services/ngalert/accesscontrol/rules.go:152]: This should remove the diff for these lines, logic is now the same as it was before.

```suggestion
	} else if len(change.Delete) > 0 {
		// add a safeguard in the case of inconsistency. If user hit this then there is a bug in the calculating of changes struct
		return fmt.Errorf("failed to authorize changes in rule group %s. Detected %d deletes but group was not provided", change.GroupKey.RuleGroup, len(change.Delete))
	}
```

File: pkg/services/ngalert/api/authorization.go [hunk 2]
```diff
@@ -20,28 +20,43 @@ func (api *API) authorize(method, path string) web.Handler {
...
-			ac.EvalPermission(ac.ActionAlertingRuleCreate, scope),
-			ac.EvalPermission(ac.ActionAlertingRuleDelete, scope),
+		eval = ac.EvalAll(
+			ac.EvalAny(
+				ac.EvalPermission(ac.ActionAlertingRuleUpdate, scope),
+				ac.EvalPermission(ac.ActionAlertingRuleCreate, scope),
+				ac.EvalPermission(ac.ActionAlertingRuleDelete, scope),
+			),
+			ac.EvalPermission(ac.ActionAlertingRuleRead, scope),
+			ac.EvalPermission(dashboards.ActionFoldersRead, scope),
```
Comment (rwwiv) [inline pkg/services/ngalert/api/authorization.go:59]: This should make the error message a bit easier to understand

```suggestion
			ac.EvalPermission(ac.ActionAlertingRuleRead, scope),
			ac.EvalPermission(dashboards.ActionFoldersRead, scope),
			ac.EvalAny(
				ac.EvalPermission(ac.ActionAlertingRuleUpdate, scope),
				ac.EvalPermission(ac.ActionAlertingRuleCreate, scope),
				ac.EvalPermission(ac.ActionAlertingRuleDelete, scope),
			),
```

File: pkg/tests/api/alerting/api_alertmanager_test.go [hunk 3]
```diff
@@ -614,7 +614,7 @@ func TestIntegrationRulerAccess(t *testing.T) {
 			desc:            "viewer request should fail",
 			client:          newAlertingApiClient(grafanaListedAddr, "viewer", "viewer"),
 			expStatus:       http.StatusForbidden,
-			expectedMessage: `You'll need additional permissions to perform this action. Permissions needed: any of alert.rules:write, alert.rules:create, alert.rules:delete`,
+			expectedMessage: `You'll need additional permissions to perform this action. Permissions needed: all of any of alert.rules:write, alert.rules:create, alert.rules:delete, alert.rules:read, folders:read`,
```
Comment (rwwiv) [inline pkg/tests/api/alerting/api_alertmanager_test.go:617]: I believe this is what the error message should be with the reordered permissions in the other suggestion.

```suggestion
			expectedMessage: `You'll need additional permissions to perform this action. Permissions needed: all of alert.rules:read, folders:read, any of alert.rules:write, alert.rules:create, alert.rules:delete`,
```

**Review by rwwiv — APPROVED** [review]
LGTM, couple minor nits with suggestions you should be able to just apply if you feel they make sense.

**Pull request merged**